Tasks/Maven/maven2.js

```diff
@@ -48,40 +51,149 @@ var publishJUnitResults = tl.getInput('publishJUnitResults');
+    if (!tl.getEndpointUrl) {
+        throw new Error(errorMessage);
+    }
+
+    var genericEndpoint = tl.getInput(inputFieldName);
+    if (!genericEndpoint) {
+        throw new Error(errorMessage);
+    }
+
+    hostUrl = tl.getEndpointUrl(genericEndpoint, false);
+    var auth = tl.getEndpointAuthorization(genericEndpoint, false);
+
+    if (auth.scheme != "UsernamePassword") {
+        throw new Error("The authorization scheme " + auth.scheme + " is not supported for a SonarQube endpoint. Please use a username and a password.");
+    }
+
+    hostUsername = auth.parameters.Username;
+    hostPassword = auth.parameters.Password;
+
+    return {
+        "Url": hostUrl,
+        "Username": hostUsername,
+        "Password": hostPassword
+    };
+}
+
+function createSonarQubeArgs(sqHostUrl, sqHostUsername, sqHostPassword, sqDbUrl, sqDbUsername, sqDbPassword) {
+    var sqArgs = "";
+    if (sqHostUrl) {
+        sqArgs += '-Dsonar.host.url="' + sqHostUrl + '" '
+    }
+    if (sqHostUsername) {
+        sqArgs += '-Dsonar.login="' + sqHostUsername + '" '
+    }
+    if (sqHostPassword) {
+        sqArgs += '-Dsonar.password="' + sqHostPassword + '" '
+    }
+    if (sqDbUrl) {
+        sqArgs += '-Dsonar.jdbc.url="' + sqDbUrl + '" '
+    }
+    if (sqDbUsername) {
+        sqArgs += '-Dsonar.jdbc.username="' + sqDbUsername + '" '
+    }
+    if (sqDbPassword) {
+        sqArgs += '-Dsonar.jdbc.password="' + sqDbPassword + '" '
+    }
+
+    return sqArgs;
+}
+
+// Maven task orchestration
+
+var runFailed = false;
+
 mvnv.exec()
-.then(function(code) {
-	return mvnb.exec();
+.then(function (code) {
+    return mvnb.exec(); // run Maven with the user specified goals
 })
-.then(function(code) {
-  publishTestResults(publishJUnitResults, testResultsFiles);
-  tl.exit(code);
+.fail(function (err) {
+    console.error(err.message);
+    runFailed = true; // record the error, but do not exit
 })
-.fail(function(err) {
-  console.error(err.message);
-  publishTestResults(publishJUnitResults, testResultsFiles);
-  tl.debug('taskRunner fail');
-  tl.exit(1);
+.fin(function () {
+    publishTestResults(publishJUnitResults, testResultsFiles); // publish test results even if tests fail, causing Maven to fail
 })
+.then(function (code) {
+    mvnsq = getSonarQubeRunner();
+
+    if (mvnsq) {
+        // run Maven with the sonar:sonar goal, even if the user-goal Maven failed (e.g. test failures)
+        // note that running sonar:sonar along with the user goals is not supported due to a SonarQube bug
+        return mvnsq.exec()  
+    }
+})
+.fail(function (err) {
+    console.error(err.message);
+    console.error("SonarQube analysis failed");
+    tl.exit(1)
+})
+.then(function (code) {
+    if (runFailed) {
+        tl.exit(1); // exit with a non-zero code to mark the entire task as having failed
+    } else {
+        tl.exit(code);
+    }
```

Comment:
If the main Maven run fails we should still be running the SQ analysis. For example if tests fail we should upload this to SQ.